src/backend/distributed/commands/view.c

```diff
@@ -0,0 +1,349 @@
+/*-------------------------------------------------------------------------
+ *
+ * view.c
+ *    Commands for distributing CREATE OR REPLACE VIEW statements.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "fmgr.h"
+
+#include "access/genam.h"
+#include "catalog/objectaddress.h"
+#include "commands/extension.h"
+#include "distributed/commands.h"
+#include "distributed/citus_ruleutils.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/deparser.h"
+#include "distributed/errormessage.h"
+#include "distributed/listutils.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/metadata/dependency.h"
+#include "distributed/metadata/distobject.h"
+#include "distributed/multi_executor.h"
+#include "distributed/namespace_utils.h"
+#include "distributed/worker_transaction.h"
+#include "executor/spi.h"
+#include "nodes/nodes.h"
+#include "nodes/pg_list.h"
+#include "tcop/utility.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
+#include "utils/syscache.h"
+
+static List * FilterNameListForDistributedViews(List *viewNamesList, bool missing_ok);
+static void AppendAliasesToCreateViewCommandForExistingView(StringInfo createViewCommand,
+															Oid viewOid);
+static void AddOptionsToCreateViewCommandForExistingView(StringInfo createViewCommand,
+														 Oid viewOid);
+
+/*
+ * PreprocessViewStmt is called during the planning phase for CREATE OR REPLACE VIEW
+ * before it is created on the local node internally.
+ */
+List *
+PreprocessViewStmt(Node *node, const char *queryString,
+				   ProcessUtilityContext processUtilityContext)
+{
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	/* check creation against multi-statement transaction policy */
+	if (!ShouldPropagateCreateInCoordinatedTransction())
```

Comment:
maybe one question to @thanodnl -- not for this PR, a general question -- should we move `ShouldPropagateCreateInCoordinatedTransction()` into `ShouldPropagate()` ?

Reply:
never mind, ShouldPropagate is called from drop/alter as well

Reply:
Alter statements mostly call `ShouldPropagate` via a function that takes the `ObjectAddress`, as a check to see if for a given object we should propagate the statements.

We could maybe make a generic should propagate create function that would check the GUC's and the transaction details. Although it think it will be worth more to deduplicate most of the create code into 1-2 different function sets that can be used to propagate the create of any object. Which is what I am working on in https://github.com/citusdata/citus/compare/refactor/dedupe-object-propagation

Reply:
btw, @thanodnl the branch looks great, hope we can land that soon ;)